### PR TITLE
client: add WithDialContext() and client.Dialer()

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

**- What I did**

Allow specifying custom dialer for hijacking

**- How I did it**

* `WithDialContext()` allows specifying custom dialer for hijacking and supposed to replace `WithDialer()`. WithDialer() is also updated to use `WithDialContext()`.
* `client.Dialer()` returns the dialer configured with `WithDialContext()`.

This is expected to be used by custom connection helpers. https://github.com/docker/cli/pull/889

**- How to verify it**

https://github.com/docker/cli/pull/889

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

image
http://www.antarctica.gov.au/about-antarctica/wildlife/animals/penguins/adelie-penguins

### CODE REVIEW

**Review by boaz0 — APPROVED** [review]
LGTM :+1: 
Can you add unit tests, though?

File: client/client.go [hunk 1]
```diff
@@ -266,7 +285,11 @@ func NewClientWithOpts(ops ...func(*Client) error) (*Client, error) {
 		// this to avoid breaking existing clients but this should be addressed.
 		c.scheme = "https"
 	}
-
+	if c.hijackDialer == nil {
+		c.hijackDialer = func(ctx context.Context, network, addr string) (net.Conn, error) {
```
Comment (dnephin) [inline client/client.go]: We should put all the defaults in the same place. All the other defaults are at the top of this function before running ops.

File: client/client.go [hunk 2]
```diff
@@ -78,6 +81,9 @@ type Client struct {
 	basePath string
 	// client used to send and receive http requests.
 	client *http.Client
+	// dialer used for hijacking. When set to non-nil, it typically corresponds to
+	// client.Transport.(*http.Tranport).DialContext .
```
Comment (dnephin) [inline client/client.go]: "When set to non-nil, it typically corresponds to client.Transport.(*http.Tranport).DialContext" seems superfluous.  Better to document appropriate values on the exported method that allows someone to set the value.

File: client/client.go [hunk 3]
```diff
@@ -209,7 +215,8 @@ func WithHost(host string) func(*Client) error {
 	}
 }
 
-// WithHTTPClient overrides the client http client with the specified one
+// WithHTTPClient overrides the client http client with the specified one.
+// See also WithHijackDialer.
```
Comment (dnephin) [inline client/client.go]: Why "See also WithHijackDialer" ? These don't seem to be related.

File: client/client.go [hunk 4]
```diff
@@ -219,6 +226,17 @@ func WithHTTPClient(client *http.Client) func(*Client) error {
 	}
 }
 
+// WithHijackDialer overrides the client default hijack dialer.
+// The typical usecase of WithHijackDialer is to allow implementing connection helpers.
```
Comment (dnephin) [inline client/client.go]: What is a connection helper? 

Isn't the "typical" use for a HijackDialer to support attach/exec ?

File: client/client.go [hunk 5]
```diff
@@ -219,6 +226,17 @@ func WithHTTPClient(client *http.Client) func(*Client) error {
...
+// The hijack dialer typically corresponds to *http.Client.Transport.(*http.Tranport).DialContext .
+// See WithHTTPClient.
```
Comment (dnephin) [inline client/client.go]: Instead of corresponds to, maybe "See http.Transport.DialContext for more information" ?

**Comment by AkihiroSuda** [comment]
Addressed the comments.

UT will be covered on CLI-side.

**Comment by AkihiroSuda** [comment]
Also added `DialRaw()`, which is going to be needed by `docker system dial-stdio` (https://github.com/docker/cli/pull/1014)

cc @tonistiigi @cpuguy83 @thaJeztah

**Comment by thaJeztah** [comment]
discussing this with @tonistiigi - let's mark this one as "WIP" until https://github.com/docker/cli/pull/1014 is moving / reviewed. SGTY @AkihiroSuda ?

**Comment by AkihiroSuda** [comment]
updated

**Comment by tonistiigi** [comment]
LGTM

cc @cpuguy83

**Review by cpuguy83 — APPROVED** [review]
LGTM
